node/request.js

```diff
@@ -83,14 +80,20 @@ TChannelRequest.defaultTimeout = 5000;
 
 TChannelRequest.prototype.type = 'tchannel.request';
 
-TChannelRequest.prototype.onError = function onError(err, self) {
+TChannelRequest.prototype.emitError = function emitError(err) {
+    var self = this;
     if (!self.end) self.end = self.timers.now();
     self.err = err;
+    self.services.onRequestError(self);
+    self.errorEvent.emit(self, err);
 };
 
-TChannelRequest.prototype.onResponse = function onResponse(res, self) {
+TChannelRequest.prototype.emitResponse = function emitResponse(res) {
+    var self = this;
     if (!self.end) self.end = self.timers.now();
     self.res = res;
+    self.services.onRequestResponse(self);
+    self.responseEvent.emit(self, res);
```

Comment:
This is exactly the change I was looking for. Thanks. (@jcorbin, the motivating principle is that we should avoid using event listeners to communicate within an instance, since we can just use a method, avoid a closure, try to avoid more than one listener on an event)